{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scheduled Integration of ClinVar Gene Variant-Disease Data into WikiData\n",
    "\n",
    "ClinVar aggregates information about genomic variation and its relationship to human health <br>\n",
    "CC0 https://www.ncbi.nlm.nih.gov/clinvar/\n",
    "\n",
    "This scheduled bot operates through WDI to integrate ClinVar Gene Variant-Disease Data <br>\n",
    "https://github.com/SuLab/GeneWikiCentral/issues/50 <br>\n",
    "\n",
    "Python script contributions, in order: Sabah Ul-Hasan, Andrew I Su"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>191195</td>\n",
       "      <td>duplication</td>\n",
       "      <td>NM_006920.6(SCN1A):c.2011-13dup</td>\n",
       "      <td>6323</td>\n",
       "      <td>SCN1A</td>\n",
       "      <td>HGNC:10585</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "      <td>0</td>\n",
       "      <td>Feb 15, 2019</td>\n",
       "      <td>549232924</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>2q24.3</td>\n",
       "      <td>criteria provided, conflicting interpretations</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>194032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>191196</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001182.5(ALDH7A1):c.1093+1G&gt;A</td>\n",
       "      <td>501</td>\n",
       "      <td>ALDH7A1</td>\n",
       "      <td>HGNC:877</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Feb 23, 2015</td>\n",
       "      <td>794727058</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5q23.2</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>194033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>191196</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001182.5(ALDH7A1):c.1093+1G&gt;A</td>\n",
       "      <td>501</td>\n",
       "      <td>ALDH7A1</td>\n",
       "      <td>HGNC:877</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Feb 23, 2015</td>\n",
       "      <td>794727058</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>5q23.2</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>194033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>191197</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001195263.2(PDZD7):c.1752T&gt;C (p.Tyr584=)</td>\n",
       "      <td>79955</td>\n",
       "      <td>PDZD7</td>\n",
       "      <td>HGNC:26257</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>Mar 16, 2015</td>\n",
       "      <td>368563439</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>10q24.31</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>194034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>191197</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001195263.2(PDZD7):c.1752T&gt;C (p.Tyr584=)</td>\n",
       "      <td>79955</td>\n",
       "      <td>PDZD7</td>\n",
       "      <td>HGNC:26257</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>Mar 16, 2015</td>\n",
       "      <td>368563439</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>10q24.31</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>194034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0     191195                duplication   \n",
       "1     191196  single nucleotide variant   \n",
       "2     191196  single nucleotide variant   \n",
       "3     191197  single nucleotide variant   \n",
       "4     191197  single nucleotide variant   \n",
       "\n",
       "                                          Name  GeneID GeneSymbol     HGNC_ID  \\\n",
       "0              NM_006920.6(SCN1A):c.2011-13dup    6323      SCN1A  HGNC:10585   \n",
       "1             NM_001182.5(ALDH7A1):c.1093+1G>A     501    ALDH7A1    HGNC:877   \n",
       "2             NM_001182.5(ALDH7A1):c.1093+1G>A     501    ALDH7A1    HGNC:877   \n",
       "3  NM_001195263.2(PDZD7):c.1752T>C (p.Tyr584=)   79955      PDZD7  HGNC:26257   \n",
       "4  NM_001195263.2(PDZD7):c.1752T>C (p.Tyr584=)   79955      PDZD7  HGNC:26257   \n",
       "\n",
       "                           ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
       "0  Conflicting interpretations of pathogenicity              0  Feb 15, 2019   \n",
       "1                                    Pathogenic              1  Feb 23, 2015   \n",
       "2                                    Pathogenic              1  Feb 23, 2015   \n",
       "3                                 Likely benign              0  Mar 16, 2015   \n",
       "4                                 Likely benign              0  Mar 16, 2015   \n",
       "\n",
       "   RS# (dbSNP)  ... ReferenceAllele AlternateAllele Cytogenetic  \\\n",
       "0    549232924  ...               T              TA      2q24.3   \n",
       "1    794727058  ...               C               T      5q23.2   \n",
       "2    794727058  ...               C               T      5q23.2   \n",
       "3    368563439  ...               A               G    10q24.31   \n",
       "4    368563439  ...               A               G    10q24.31   \n",
       "\n",
       "                                     ReviewStatus NumberSubmitters Guidelines  \\\n",
       "0  criteria provided, conflicting interpretations                3        NaN   \n",
       "1             criteria provided, single submitter                1        NaN   \n",
       "2             criteria provided, single submitter                1        NaN   \n",
       "3             criteria provided, single submitter                1        NaN   \n",
       "4             criteria provided, single submitter                1        NaN   \n",
       "\n",
       "  TestedInGTR OtherIDs SubmitterCategories  VariationID  \n",
       "0           N        -                   2       194032  \n",
       "1           N        -                   2       194033  \n",
       "2           N        -                   2       194033  \n",
       "3           N        -                   2       194034  \n",
       "4           N        -                   2       194034  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download data from NCBI\n",
    "\n",
    "# Make sure os has wget installed, or the command wont work\n",
    "import os\n",
    "os.system('wget ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz') \n",
    "\n",
    "# Create time stamp of when downloaded (error if isoformat() used)\n",
    "from datetime import datetime\n",
    "timeStringNow = datetime.now().strftime(\"+%Y-%m-%dT00:00:00Z\")\n",
    "\n",
    "# Unzip the file\n",
    "import gzip\n",
    "import shutil\n",
    "with gzip.open('variant_summary.txt.gz', 'rb') as f_in:\n",
    "    with open('variant_summary.txt', 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "        \n",
    "# Convert .txt file to .csv\n",
    "import pandas as pd \n",
    "import csv\n",
    "\n",
    "txt_file = r\"variant_summary.txt\"\n",
    "csv_file = r\"variant_summary.csv\"\n",
    "\n",
    "with open(txt_file, \"r\") as in_text:\n",
    "    in_reader = csv.reader(in_text, delimiter = '\\t')\n",
    "    with open(csv_file, \"w\") as out_csv:\n",
    "        out_writer = csv.writer(out_csv)\n",
    "        for row in in_reader:\n",
    "            out_writer.writerow(row)\n",
    "\n",
    "# Import .csv file and read first 5 rows\n",
    "data = pd.read_csv(\"variant_summary.csv\") \n",
    "data.shape # 31 columns, 1319815 rows\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Start</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>ReviewStatus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>duplication</td>\n",
       "      <td>NM_006920.6(SCN1A):c.2011-13dup</td>\n",
       "      <td>6323</td>\n",
       "      <td>SCN1A</td>\n",
       "      <td>HGNC:10585</td>\n",
       "      <td>Conflicting interpretations of pathogenicity</td>\n",
       "      <td>194032</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>166042429</td>\n",
       "      <td>T</td>\n",
       "      <td>TA</td>\n",
       "      <td>criteria provided, conflicting interpretations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001182.5(ALDH7A1):c.1093+1G&gt;A</td>\n",
       "      <td>501</td>\n",
       "      <td>ALDH7A1</td>\n",
       "      <td>HGNC:877</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>194033</td>\n",
       "      <td>NC_000005.9</td>\n",
       "      <td>125891622</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001182.5(ALDH7A1):c.1093+1G&gt;A</td>\n",
       "      <td>501</td>\n",
       "      <td>ALDH7A1</td>\n",
       "      <td>HGNC:877</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>194033</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>126555930</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001195263.2(PDZD7):c.1752T&gt;C (p.Tyr584=)</td>\n",
       "      <td>79955</td>\n",
       "      <td>PDZD7</td>\n",
       "      <td>HGNC:26257</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>194034</td>\n",
       "      <td>NC_000010.10</td>\n",
       "      <td>102772013</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001195263.2(PDZD7):c.1752T&gt;C (p.Tyr584=)</td>\n",
       "      <td>79955</td>\n",
       "      <td>PDZD7</td>\n",
       "      <td>HGNC:26257</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>194034</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>101012256</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Type                                         Name  \\\n",
       "0                duplication              NM_006920.6(SCN1A):c.2011-13dup   \n",
       "1  single nucleotide variant             NM_001182.5(ALDH7A1):c.1093+1G>A   \n",
       "2  single nucleotide variant             NM_001182.5(ALDH7A1):c.1093+1G>A   \n",
       "3  single nucleotide variant  NM_001195263.2(PDZD7):c.1752T>C (p.Tyr584=)   \n",
       "4  single nucleotide variant  NM_001195263.2(PDZD7):c.1752T>C (p.Tyr584=)   \n",
       "\n",
       "   GeneID GeneSymbol     HGNC_ID  \\\n",
       "0    6323      SCN1A  HGNC:10585   \n",
       "1     501    ALDH7A1    HGNC:877   \n",
       "2     501    ALDH7A1    HGNC:877   \n",
       "3   79955      PDZD7  HGNC:26257   \n",
       "4   79955      PDZD7  HGNC:26257   \n",
       "\n",
       "                           ClinicalSignificance  VariationID  \\\n",
       "0  Conflicting interpretations of pathogenicity       194032   \n",
       "1                                    Pathogenic       194033   \n",
       "2                                    Pathogenic       194033   \n",
       "3                                 Likely benign       194034   \n",
       "4                                 Likely benign       194034   \n",
       "\n",
       "  ChromosomeAccession      Start ReferenceAllele AlternateAllele  \\\n",
       "0        NC_000002.12  166042429               T              TA   \n",
       "1         NC_000005.9  125891622               C               T   \n",
       "2        NC_000005.10  126555930               C               T   \n",
       "3        NC_000010.10  102772013               A               G   \n",
       "4        NC_000010.11  101012256               A               G   \n",
       "\n",
       "                                     ReviewStatus  \n",
       "0  criteria provided, conflicting interpretations  \n",
       "1             criteria provided, single submitter  \n",
       "2             criteria provided, single submitter  \n",
       "3             criteria provided, single submitter  \n",
       "4             criteria provided, single submitter  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean-up the data for identification and integration\n",
    "\n",
    "# Columns to keep\n",
    "new = data[['Type', 'Name','GeneID','GeneSymbol','HGNC_ID', 'ClinicalSignificance','VariationID', \n",
    "            'ChromosomeAccession', 'Start', 'ReferenceAllele', 'AlternateAllele',\n",
    "            'ReviewStatus']]\n",
    "new.shape # 12 columns, 1319815 rows\n",
    "new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  after removing the cwd from sys.path.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(38, 13)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create new column that converts 'ReviewStatus' to star rating\n",
    "## https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/\n",
    "\n",
    "new['Rating'] = \"\" # Create empty column for gold star rating\n",
    "\n",
    "## Convert strings from 'ReviewStatus' to 'Rating' \n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion provided'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion criteria provided'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion for the individual variant'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, single submitter'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, conflicting interpretations'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, conflicting interpretations'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, multiple submitters, no conflicts'), 'Rating'] = 'two'\n",
    "new.loc[new['ReviewStatus'].str.contains('reviewed by expert panel'), 'Rating'] = 'three'\n",
    "new.loc[new['ReviewStatus'].str.contains('practice guideline'), 'Rating'] = 'four'\n",
    "\n",
    "## Create a report that automatically outputs a summary of these each download?\n",
    "# Keep anything with 'four' or more stars in 'Rating' column\n",
    "four = new[new['Rating'].str.contains('four')] # excludes both one and none\n",
    "four.shape # 13 columns, 63 rows\n",
    "# Keep anything noted as 'single nucleotide variant' in 'Type' column\n",
    "snv = four[four['Type'].str.contains('single nucleotide variant')]\n",
    "snv.shape # 54 rows\n",
    "# Keep anything with 'Pathogenic' in the 'ClinicalSignificance' column\n",
    "patho = snv[snv['ClinicalSignificance'].str.contains('Pathogenic')]\n",
    "patho.shape # 38 rows "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  \"\"\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Start</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Rating</th>\n",
       "      <th>HGVS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001369.2(DNAH5):c.1585C&gt;T (p.Arg529Trp)</td>\n",
       "      <td>1767</td>\n",
       "      <td>DNAH5</td>\n",
       "      <td>HGNC:2950</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>194035</td>\n",
       "      <td>NC_000005.9</td>\n",
       "      <td>13911554</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>two</td>\n",
       "      <td>NC_000005.9:g.13911554G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001369.2(DNAH5):c.1585C&gt;T (p.Arg529Trp)</td>\n",
       "      <td>1767</td>\n",
       "      <td>DNAH5</td>\n",
       "      <td>HGNC:2950</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>194035</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>13911445</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>two</td>\n",
       "      <td>NC_000005.10:g.13911445G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001429.4(EP300):c.2240C&gt;T (p.Pro747Leu)</td>\n",
       "      <td>2033</td>\n",
       "      <td>EP300</td>\n",
       "      <td>HGNC:3373</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>194036</td>\n",
       "      <td>NC_000022.10</td>\n",
       "      <td>41543949</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>two</td>\n",
       "      <td>NC_000022.10:g.41543949C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001429.4(EP300):c.2240C&gt;T (p.Pro747Leu)</td>\n",
       "      <td>2033</td>\n",
       "      <td>EP300</td>\n",
       "      <td>HGNC:3373</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>194036</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>41147945</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>two</td>\n",
       "      <td>NC_000022.11:g.41147945C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001458.4(FLNC):c.1902G&gt;A (p.Glu634=)</td>\n",
       "      <td>2318</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>HGNC:3756</td>\n",
       "      <td>Benign</td>\n",
       "      <td>194038</td>\n",
       "      <td>NC_000007.13</td>\n",
       "      <td>128481312</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>criteria provided, multiple submitters, no con...</td>\n",
       "      <td>two</td>\n",
       "      <td>NC_000007.13:g.128481312G&gt;A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Type                                        Name  \\\n",
       "5   single nucleotide variant  NM_001369.2(DNAH5):c.1585C>T (p.Arg529Trp)   \n",
       "6   single nucleotide variant  NM_001369.2(DNAH5):c.1585C>T (p.Arg529Trp)   \n",
       "7   single nucleotide variant  NM_001429.4(EP300):c.2240C>T (p.Pro747Leu)   \n",
       "8   single nucleotide variant  NM_001429.4(EP300):c.2240C>T (p.Pro747Leu)   \n",
       "11  single nucleotide variant     NM_001458.4(FLNC):c.1902G>A (p.Glu634=)   \n",
       "\n",
       "    GeneID GeneSymbol    HGNC_ID    ClinicalSignificance  VariationID  \\\n",
       "5     1767      DNAH5  HGNC:2950  Uncertain significance       194035   \n",
       "6     1767      DNAH5  HGNC:2950  Uncertain significance       194035   \n",
       "7     2033      EP300  HGNC:3373    Benign/Likely benign       194036   \n",
       "8     2033      EP300  HGNC:3373    Benign/Likely benign       194036   \n",
       "11    2318       FLNC  HGNC:3756                  Benign       194038   \n",
       "\n",
       "   ChromosomeAccession      Start ReferenceAllele AlternateAllele  \\\n",
       "5          NC_000005.9   13911554               G               A   \n",
       "6         NC_000005.10   13911445               G               A   \n",
       "7         NC_000022.10   41543949               C               T   \n",
       "8         NC_000022.11   41147945               C               T   \n",
       "11        NC_000007.13  128481312               G               A   \n",
       "\n",
       "                                         ReviewStatus Rating  \\\n",
       "5   criteria provided, multiple submitters, no con...    two   \n",
       "6   criteria provided, multiple submitters, no con...    two   \n",
       "7   criteria provided, multiple submitters, no con...    two   \n",
       "8   criteria provided, multiple submitters, no con...    two   \n",
       "11  criteria provided, multiple submitters, no con...    two   \n",
       "\n",
       "                           HGVS  \n",
       "5     NC_000005.9:g.13911554G>A  \n",
       "6    NC_000005.10:g.13911445G>A  \n",
       "7    NC_000022.10:g.41543949C>T  \n",
       "8    NC_000022.11:g.41147945C>T  \n",
       "11  NC_000007.13:g.128481312G>A  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create HGVS Column and IDs\n",
    "snv['HGVS'] = \"\" # Create empty column for HGVS nomenclature\n",
    "\n",
    "test = snv\n",
    "test['HGVS']=test['ChromosomeAccession']+':g.'+test['Start'].astype(str)+test['ReferenceAllele']+'>'+test['AlternateAllele']\n",
    "test.head()\n",
    "# NC_000007.13:g.140453136A>T\n",
    "\n",
    "# ChromosomeAccession \n",
    "# NC_000007.13\n",
    "\n",
    "# Start\n",
    "# 140453136\n",
    "\n",
    "#ReferenceAllele\n",
    "# A\n",
    "\n",
    "#AlternateAllele\n",
    "# T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Jan 6\n",
    "# Keep anything with 'two' or more stars in 'Rating' column\n",
    "twoplus = new[~new['Rating'].str.contains('one')] # excludes both one and none\n",
    "twoplus.shape # 13 columns, 196530 rows: 1123285 removed, 14.9 % of all data usable\n",
    "# Keep anything noted as 'single nucleotide variant' in 'Type' column\n",
    "snv = twoplus[twoplus['Type'].str.contains('single nucleotide variant')]\n",
    "snv.shape # 174733 rows, or 88.9% of data with two or more stars (13.2% of all data)\n",
    "## 83.2 % of total data, prior to star rating filter, are SNVs\n",
    "# Keep anything with 'Pathogenic' in the 'ClinicalSignificance' column\n",
    "patho = snv[snv['ClinicalSignificance'].str.contains('Pathogenic')]\n",
    "patho.shape # 20040 rows, or 10.2% of data with two or more stars that are snvs (1.5% of all data)\n",
    "## 12.8 % of total data, prior to star rating filter, are Pathogenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already up-to-date: pip in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (19.3.1)\n",
      "Requirement already satisfied: tqdm in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (4.32.1)\n",
      "Requirement already satisfied: termcolor in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (1.1.0)\n",
      "Requirement already satisfied: wikidataintegrator in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (0.4.2)\n",
      "Requirement already satisfied: python-dateutil in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (2.8.0)\n",
      "Requirement already satisfied: requests in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (2.22.0)\n",
      "Requirement already satisfied: ShExJSG in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (0.6.8)\n",
      "Requirement already satisfied: oauthlib in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (3.1.0)\n",
      "Requirement already satisfied: jsonasobj in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (1.2.1)\n",
      "Requirement already satisfied: pyshex in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (0.7.12)\n",
      "Requirement already satisfied: sparql-slurper in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (0.2.2)\n",
      "Requirement already satisfied: simplejson in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (3.16.0)\n",
      "Requirement already satisfied: mwoauth in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from wikidataintegrator) (0.3.5)\n",
      "Requirement already satisfied: six>=1.5 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from python-dateutil->wikidataintegrator) (1.12.0)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from requests->wikidataintegrator) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from requests->wikidataintegrator) (2019.6.16)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from requests->wikidataintegrator) (1.24.2)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from requests->wikidataintegrator) (2.8)\n",
      "Requirement already satisfied: pyshexc>=0.5.4 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from ShExJSG->wikidataintegrator) (0.7.0)\n",
      "Requirement already satisfied: pyjsg>=0.9.0 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from ShExJSG->wikidataintegrator) (0.9.1)\n",
      "Requirement already satisfied: sparqlwrapper in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from pyshex->wikidataintegrator) (1.8.4)\n",
      "Requirement already satisfied: rdflib>=4.2.2 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from pyshex->wikidataintegrator) (4.2.2)\n",
      "Requirement already satisfied: cfgraph>=0.2.1 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from pyshex->wikidataintegrator) (0.2.1)\n",
      "Requirement already satisfied: rdflib-jsonld>=0.4.0 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from pyshex->wikidataintegrator) (0.4.0)\n",
      "Requirement already satisfied: requests-oauthlib in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from mwoauth->wikidataintegrator) (1.2.0)\n",
      "Requirement already satisfied: PyJWT<2.0.0,>=1.0.1 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from mwoauth->wikidataintegrator) (1.7.1)\n",
      "Requirement already satisfied: antlr4-python3-runtime>=4.7 in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from pyshexc>=0.5.4->ShExJSG->wikidataintegrator) (4.7.2)\n",
      "Requirement already satisfied: pyparsing in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from rdflib>=4.2.2->pyshex->wikidataintegrator) (2.4.0)\n",
      "Requirement already satisfied: isodate in /Users/sulhasan/anaconda3/lib/python3.7/site-packages (from rdflib>=4.2.2->pyshex->wikidataintegrator) (0.6.0)\n"
     ]
    }
   ],
   "source": [
    "# Installations by shell \n",
    "!pip install --upgrade pip # Installs pip, ensures it's up-to-date\n",
    "!pip install tqdm # Visualizes installation progress (progress bar)\n",
    "!pip install termcolor # For color-coding printed output\n",
    "!pip install wikidataintegrator # For wikidata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Installations by python\n",
    "from wikidataintegrator import wdi_core, wdi_login # Core and login from wikidataintegrator module\n",
    "from wikidataintegrator.ref_handlers import update_retrieved_if_new_multiple_refs # For retrieving references\n",
    "import copy # Copies references needed in the .csv for uploading to wikidata\n",
    "from datetime import datetime # For identifying the current date and time\n",
    "import time # For keeping track of total for loop run time\n",
    "\n",
    "import os # OS package to ensure interaction between the modules (ie WDI) and current OS being used\n",
    "\n",
    "import pandas as pd # Pandas for data organization, then abbreviated to pd\n",
    "import numpy as np # Another general purpose package"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ClinVar Gen gene-disease validity data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "source": [
    "# Automate this from ftp file later...\n",
    "import urllib \n",
    "\n",
    "file = urllib.request.urlopen('ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz', 'variant_summary.txt.gz')\n",
    "\n",
    "with open('variant_summary.txt') as f:\n",
    "    mylist = list(f)\n",
    "\n",
    "print(mylist)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Login for running WDI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logging in...\n",
      "https://www.wikidata.org/w/api.php\n",
      "Successfully logged in as Sulhasan\n"
     ]
    }
   ],
   "source": [
    "print(\"Logging in...\") \n",
    "\n",
    "# Enter your own username and password ** to be updated to ProteinBoxBot\n",
    "os.environ[\"WDUSER\"] = \"username\" # Uses os package to call and set the environment for wikidata username\n",
    "os.environ[\"WDPASS\"] = \"password\"\n",
    "\n",
    "# Conditional that outputs error command if not in the local python environment\n",
    "if \"WDUSER\" in os.environ and \"WDPASS\" in os.environ: \n",
    "    WDUSER = os.environ['WDUSER']\n",
    "    WDPASS = os.environ['WDPASS']\n",
    "else: \n",
    "    raise ValueError(\"WDUSER and WDPASS must be specified in local.py or as environment variables\")      \n",
    "\n",
    "# Sets attributed username and password as 'login'\n",
    "login = wdi_login.WDLogin(WDUSER, WDPASS) "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
