{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scheduled Integration of ClinVar Gene Variant-Disease Data into WikiData\n",
    "\n",
    "ClinVar aggregates information about genomic variation and its relationship to human health <br>\n",
    "CC0 https://www.ncbi.nlm.nih.gov/clinvar/\n",
    "\n",
    "This scheduled bot operates monthly through WDI to integrate ClinVar Gene Variant-Disease Data <br>\n",
    "https://www.ncbi.nlm.nih.gov/clinvar/docs/ftp_primer/ (variant_summary) <br>\n",
    "https://github.com/SuLab/GeneWikiCentral/issues/50 <br>\n",
    "http://jenkins.sulab.org/ <br>\n",
    "\n",
    "Python script contributions, in order: Sabah Ul-Hasan, Andrew I Su, Tong Shu Li"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Checks and Tests\n",
    "\n",
    "- x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relevant Modules and Libraries\n",
    "\n",
    "import os # OS package to ensure interaction between the modules (ie WDI) and current OS being used\n",
    "\n",
    "from datetime import datetime # For identifying the current date and time\n",
    "import time # Keep track of total for loop run time\n",
    "\n",
    "import gzip # For unzip of files\n",
    "import shutil # Copies content of source file(s)\n",
    "import csv # For converting file(s) to csv format\n",
    "\n",
    "import pandas as pd # For data organization, abbreviated to pd\n",
    "import numpy as np # For data organization, abbreviated as np\n",
    "\n",
    "from wikidataintegrator import wdi_core, wdi_login # Core and login from wikidataintegrator module\n",
    "from wikidataintegrator.ref_handlers import update_retrieved_if_new_multiple_refs # For retrieving references\n",
    "from wikidataintegrator.wdi_helpers import try_write # For writing new item pages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1339084, 31)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download data from NCBI\n",
    "\n",
    "## Make sure os has wget installed, or the following command wont work\n",
    "os.system('wget ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz') \n",
    "timeStringNow = datetime.now().strftime(\"+%Y-%m-%dT00:00:00Z\") # time stamp of download \n",
    "\n",
    "## Unzip the file\n",
    "with gzip.open('variant_summary.txt.gz', 'rb') as f_in:\n",
    "    with open('variant_summary.txt', 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "        \n",
    "## Convert .txt file to .csv\n",
    "txt_file = r\"variant_summary.txt\"\n",
    "csv_file = r\"variant_summary.csv\"\n",
    "\n",
    "with open(txt_file, \"r\") as in_text:\n",
    "    in_reader = csv.reader(in_text, delimiter = '\\t')\n",
    "    with open(csv_file, \"w\") as out_csv:\n",
    "        out_writer = csv.writer(out_csv)\n",
    "        for row in in_reader:\n",
    "            out_writer.writerow(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>156440</td>\n",
       "      <td>copy number loss</td>\n",
       "      <td>GRCh38/hg38 13q12.12(chr13:22822881-23533846)x1</td>\n",
       "      <td>-1</td>\n",
       "      <td>LINC00327;LINC00621;SACS;SACS-AS1;SGCG</td>\n",
       "      <td>-</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Feb 25, 2011</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>dbVar:nssv706495,dbVar:nsv533231</td>\n",
       "      <td>2</td>\n",
       "      <td>146689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>156440</td>\n",
       "      <td>copy number loss</td>\n",
       "      <td>GRCh38/hg38 13q12.12(chr13:22822881-23533846)x1</td>\n",
       "      <td>-1</td>\n",
       "      <td>LINC00327;LINC00621;SACS;SACS-AS1;SGCG</td>\n",
       "      <td>-</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Feb 25, 2011</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>dbVar:nssv706495,dbVar:nsv533231</td>\n",
       "      <td>2</td>\n",
       "      <td>146689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>156440</td>\n",
       "      <td>copy number loss</td>\n",
       "      <td>GRCh38/hg38 13q12.12(chr13:22822881-23533846)x1</td>\n",
       "      <td>-1</td>\n",
       "      <td>LINC00327;LINC00621;SACS;SACS-AS1;SGCG</td>\n",
       "      <td>-</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Feb 25, 2011</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>13q12.12</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>dbVar:nssv706495,dbVar:nsv533231</td>\n",
       "      <td>2</td>\n",
       "      <td>146689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>156441</td>\n",
       "      <td>copy number loss</td>\n",
       "      <td>GRCh38/hg38 18q23(chr18:79754409-79877194)x1</td>\n",
       "      <td>-1</td>\n",
       "      <td>CTDP1;KCNG2</td>\n",
       "      <td>-</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>May 06, 2011</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>18q23</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>dbVar:nssv706496,dbVar:nsv533232</td>\n",
       "      <td>2</td>\n",
       "      <td>146690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>156441</td>\n",
       "      <td>copy number loss</td>\n",
       "      <td>GRCh38/hg38 18q23(chr18:79754409-79877194)x1</td>\n",
       "      <td>-1</td>\n",
       "      <td>CTDP1;KCNG2</td>\n",
       "      <td>-</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>May 06, 2011</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>18q23</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>dbVar:nssv706496,dbVar:nsv533232</td>\n",
       "      <td>2</td>\n",
       "      <td>146690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID              Type  \\\n",
       "0     156440  copy number loss   \n",
       "1     156440  copy number loss   \n",
       "2     156440  copy number loss   \n",
       "3     156441  copy number loss   \n",
       "4     156441  copy number loss   \n",
       "\n",
       "                                              Name  GeneID  \\\n",
       "0  GRCh38/hg38 13q12.12(chr13:22822881-23533846)x1      -1   \n",
       "1  GRCh38/hg38 13q12.12(chr13:22822881-23533846)x1      -1   \n",
       "2  GRCh38/hg38 13q12.12(chr13:22822881-23533846)x1      -1   \n",
       "3     GRCh38/hg38 18q23(chr18:79754409-79877194)x1      -1   \n",
       "4     GRCh38/hg38 18q23(chr18:79754409-79877194)x1      -1   \n",
       "\n",
       "                               GeneSymbol HGNC_ID    ClinicalSignificance  \\\n",
       "0  LINC00327;LINC00621;SACS;SACS-AS1;SGCG       -  Uncertain significance   \n",
       "1  LINC00327;LINC00621;SACS;SACS-AS1;SGCG       -  Uncertain significance   \n",
       "2  LINC00327;LINC00621;SACS;SACS-AS1;SGCG       -  Uncertain significance   \n",
       "3                             CTDP1;KCNG2       -           Likely benign   \n",
       "4                             CTDP1;KCNG2       -           Likely benign   \n",
       "\n",
       "   ClinSigSimple LastEvaluated  RS# (dbSNP)  ... ReferenceAllele  \\\n",
       "0              0  Feb 25, 2011           -1  ...              na   \n",
       "1              0  Feb 25, 2011           -1  ...              na   \n",
       "2              0  Feb 25, 2011           -1  ...              na   \n",
       "3              0  May 06, 2011           -1  ...              na   \n",
       "4              0  May 06, 2011           -1  ...              na   \n",
       "\n",
       "  AlternateAllele Cytogenetic                    ReviewStatus  \\\n",
       "0              na    13q12.12  no assertion criteria provided   \n",
       "1              na    13q12.12  no assertion criteria provided   \n",
       "2              na    13q12.12  no assertion criteria provided   \n",
       "3              na       18q23  no assertion criteria provided   \n",
       "4              na       18q23  no assertion criteria provided   \n",
       "\n",
       "  NumberSubmitters Guidelines TestedInGTR                          OtherIDs  \\\n",
       "0                1        NaN           N  dbVar:nssv706495,dbVar:nsv533231   \n",
       "1                1        NaN           N  dbVar:nssv706495,dbVar:nsv533231   \n",
       "2                1        NaN           N  dbVar:nssv706495,dbVar:nsv533231   \n",
       "3                1        NaN           N  dbVar:nssv706496,dbVar:nsv533232   \n",
       "4                1        NaN           N  dbVar:nssv706496,dbVar:nsv533232   \n",
       "\n",
       "  SubmitterCategories  VariationID  \n",
       "0                   2       146689  \n",
       "1                   2       146689  \n",
       "2                   2       146689  \n",
       "3                   2       146690  \n",
       "4                   2       146690  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Import .csv file and read first 5 rows\n",
    "df = pd.read_csv(\"variant_summary.csv\") \n",
    "df.shape # 31 columns, 1339084 rows\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Rating</th>\n",
       "      <th>HGVS_NC</th>\n",
       "      <th>HGVS_chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23254</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000059.3(BRCA2):c.8490G&gt;A (p.Trp2830Ter)</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>HGNC:1101</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C2675520,OMIM:612555;MedGen:C0006142,OM...</td>\n",
       "      <td>Breast-ovarian cancer, familial 2;Familial can...</td>\n",
       "      <td>155722</td>\n",
       "      <td>NC_000013.10</td>\n",
       "      <td>13</td>\n",
       "      <td>32945095</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000013.10:g.32945095G&gt;A</td>\n",
       "      <td>chr13:g.32945095G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23255</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000059.3(BRCA2):c.8490G&gt;A (p.Trp2830Ter)</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>HGNC:1101</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C2675520,OMIM:612555;MedGen:C0006142,OM...</td>\n",
       "      <td>Breast-ovarian cancer, familial 2;Familial can...</td>\n",
       "      <td>155722</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>13</td>\n",
       "      <td>32370958</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000013.11:g.32370958G&gt;A</td>\n",
       "      <td>chr13:g.32370958G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24036</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000059.3(BRCA2):c.6044T&gt;A (p.Leu2015Ter)</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>HGNC:1101</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C2675520,OMIM:612555;MedGen:C0027672,SN...</td>\n",
       "      <td>Breast-ovarian cancer, familial 2;Hereditary c...</td>\n",
       "      <td>156172</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>13</td>\n",
       "      <td>32340399</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000013.11:g.32340399T&gt;A</td>\n",
       "      <td>chr13:g.32340399T&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24037</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000059.3(BRCA2):c.6044T&gt;A (p.Leu2015Ter)</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>HGNC:1101</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C2675520,OMIM:612555;MedGen:C0027672,SN...</td>\n",
       "      <td>Breast-ovarian cancer, familial 2;Hereditary c...</td>\n",
       "      <td>156172</td>\n",
       "      <td>NC_000013.10</td>\n",
       "      <td>13</td>\n",
       "      <td>32914536</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000013.10:g.32914536T&gt;A</td>\n",
       "      <td>chr13:g.32914536T&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24358</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_004360.5(CDH1):c.1023T&gt;G (p.Tyr341Ter)</td>\n",
       "      <td>CDH1</td>\n",
       "      <td>HGNC:1748</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C1708349,OMIM:137215,Orphanet:ORPHA26106</td>\n",
       "      <td>Hereditary diffuse gastric cancer</td>\n",
       "      <td>156374</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16</td>\n",
       "      <td>68812149</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000016.10:g.68812149T&gt;G</td>\n",
       "      <td>chr16:g.68812149T&gt;G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Type                                         Name  \\\n",
       "23254  single nucleotide variant  NM_000059.3(BRCA2):c.8490G>A (p.Trp2830Ter)   \n",
       "23255  single nucleotide variant  NM_000059.3(BRCA2):c.8490G>A (p.Trp2830Ter)   \n",
       "24036  single nucleotide variant  NM_000059.3(BRCA2):c.6044T>A (p.Leu2015Ter)   \n",
       "24037  single nucleotide variant  NM_000059.3(BRCA2):c.6044T>A (p.Leu2015Ter)   \n",
       "24358  single nucleotide variant    NM_004360.5(CDH1):c.1023T>G (p.Tyr341Ter)   \n",
       "\n",
       "      GeneSymbol    HGNC_ID ClinicalSignificance  \\\n",
       "23254      BRCA2  HGNC:1101           Pathogenic   \n",
       "23255      BRCA2  HGNC:1101           Pathogenic   \n",
       "24036      BRCA2  HGNC:1101           Pathogenic   \n",
       "24037      BRCA2  HGNC:1101           Pathogenic   \n",
       "24358       CDH1  HGNC:1748           Pathogenic   \n",
       "\n",
       "                                            PhenotypeIDS  \\\n",
       "23254  MedGen:C2675520,OMIM:612555;MedGen:C0006142,OM...   \n",
       "23255  MedGen:C2675520,OMIM:612555;MedGen:C0006142,OM...   \n",
       "24036  MedGen:C2675520,OMIM:612555;MedGen:C0027672,SN...   \n",
       "24037  MedGen:C2675520,OMIM:612555;MedGen:C0027672,SN...   \n",
       "24358    MedGen:C1708349,OMIM:137215,Orphanet:ORPHA26106   \n",
       "\n",
       "                                           PhenotypeList  VariationID  \\\n",
       "23254  Breast-ovarian cancer, familial 2;Familial can...       155722   \n",
       "23255  Breast-ovarian cancer, familial 2;Familial can...       155722   \n",
       "24036  Breast-ovarian cancer, familial 2;Hereditary c...       156172   \n",
       "24037  Breast-ovarian cancer, familial 2;Hereditary c...       156172   \n",
       "24358                  Hereditary diffuse gastric cancer       156374   \n",
       "\n",
       "      ChromosomeAccession Chromosome     Start ReferenceAllele  \\\n",
       "23254        NC_000013.10         13  32945095               G   \n",
       "23255        NC_000013.11         13  32370958               G   \n",
       "24036        NC_000013.11         13  32340399               T   \n",
       "24037        NC_000013.10         13  32914536               T   \n",
       "24358        NC_000016.10         16  68812149               T   \n",
       "\n",
       "      AlternateAllele              ReviewStatus Rating  \\\n",
       "23254               A  reviewed by expert panel  three   \n",
       "23255               A  reviewed by expert panel  three   \n",
       "24036               A  reviewed by expert panel  three   \n",
       "24037               A  reviewed by expert panel  three   \n",
       "24358               G  reviewed by expert panel  three   \n",
       "\n",
       "                          HGVS_NC             HGVS_chr  \n",
       "23254  NC_000013.10:g.32945095G>A  chr13:g.32945095G>A  \n",
       "23255  NC_000013.11:g.32370958G>A  chr13:g.32370958G>A  \n",
       "24036  NC_000013.11:g.32340399T>A  chr13:g.32340399T>A  \n",
       "24037  NC_000013.10:g.32914536T>A  chr13:g.32914536T>A  \n",
       "24358  NC_000016.10:g.68812149T>G  chr16:g.68812149T>G  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean-up of data for identification and integration\n",
    "\n",
    "## Columns to keep\n",
    "new = df[['Type', 'Name', 'GeneSymbol','HGNC_ID', 'ClinicalSignificance', 'PhenotypeIDS', 'PhenotypeList','VariationID', \n",
    "            'ChromosomeAccession', 'Chromosome', 'Start', 'ReferenceAllele', 'AlternateAllele',\n",
    "            'ReviewStatus']]\n",
    "\n",
    "## Create new column that converts 'ReviewStatus' to star rating\n",
    "### https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/\n",
    "new['Rating'] = \"\" # Create empty column for gold star rating\n",
    "## Convert strings from 'ReviewStatus' to 'Rating' \n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion provided'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion criteria provided'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion for the individual variant'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, single submitter'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, conflicting interpretations'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, conflicting interpretations'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, multiple submitters, no conflicts'), 'Rating'] = 'two'\n",
    "new.loc[new['ReviewStatus'].str.contains('reviewed by expert panel'), 'Rating'] = 'three'\n",
    "new.loc[new['ReviewStatus'].str.contains('practice guideline'), 'Rating'] = 'four'\n",
    "\n",
    "new.shape # 15 columns, 1339084 rows \n",
    "\n",
    "## Training dataset, based on the following criteria\n",
    "threeplus=new[new['Rating'].str.contains('three|four')]\n",
    "snv=threeplus[threeplus['Type'].str.contains('single nucleotide variant')]\n",
    "patho=snv[snv['ClinicalSignificance'].str.contains('Pathogenic')]\n",
    "\n",
    "patho.shape # 15 columns, 3870 rows (0.29% of original dataset)\n",
    "\n",
    "## Create HGVS columns and IDs (based on genomic position)\n",
    "### Naming nomenclature: https://varnomen.hgvs.org/bg-material/numbering/\n",
    "patho['HGVS_NC'] = \"\" # Create empty column for HGVS nomenclature with NC\n",
    "patho['HGVS_chr'] = \"\" # Create empty column for HGVS nomenclature with chr\n",
    "\n",
    "patho['HGVS_NC']=patho['ChromosomeAccession']+':g.'+patho['Start'].astype(str)+patho['ReferenceAllele']+'>'+patho['AlternateAllele']\n",
    "patho['HGVS_chr']='chr'+patho['Chromosome'].astype(str)+':g.'+patho['Start'].astype(str)+patho['ReferenceAllele']+'>'+patho['AlternateAllele']\n",
    "\n",
    "df=patho\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logging in...\n",
      "https://www.wikidata.org/w/api.php\n",
      "Successfully logged in as Sulhasan\n"
     ]
    }
   ],
   "source": [
    "# Login for running WDI\n",
    "print(\"Logging in...\") \n",
    "\n",
    "# **remove lines when scheduling to Jenkins** Enter your own username and password \n",
    "os.environ[\"WDUSER\"] = \"username\" # Uses os package to call and set the environment for wikidata username\n",
    "os.environ[\"WDPASS\"] = \"password\"\n",
    "\n",
    "## Conditional that outputs error command if not in the local python environment\n",
    "if \"WDUSER\" in os.environ and \"WDPASS\" in os.environ: \n",
    "    WDUSER = os.environ['WDUSER']\n",
    "    WDPASS = os.environ['WDPASS']\n",
    "else: \n",
    "    raise ValueError(\"WDUSER and WDPASS must be specified in local.py or as environment variables\")      \n",
    "\n",
    "## Sets attributed username and password as 'login'\n",
    "login = wdi_login.WDLogin(WDUSER, WDPASS) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Rating</th>\n",
       "      <th>HGVS_NC</th>\n",
       "      <th>HGVS_chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63376</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000314.7(PTEN):c.517C&gt;T (p.Arg173Cys)</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>HGNC:9588</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:CN072330,OMIM:158350;MedGen:CN072330,OM...</td>\n",
       "      <td>Cowden syndrome 1;Cowden syndrome 1;Glioma sus...</td>\n",
       "      <td>189500</td>\n",
       "      <td>NC_000010.10</td>\n",
       "      <td>10</td>\n",
       "      <td>89711899</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000010.10:g.89711899C&gt;T</td>\n",
       "      <td>chr10:g.89711899C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141137</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1A&gt;G (p.Met1Val)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Human Phenotype Ontology:HP:0004923,MedGen:C07...</td>\n",
       "      <td>Hyperphenylalaninemia, non-pku;Phenylketonuria...</td>\n",
       "      <td>586</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>103310908</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.11:g.103310908T&gt;C</td>\n",
       "      <td>chr12:g.103310908T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141138</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1A&gt;G (p.Met1Val)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Human Phenotype Ontology:HP:0004923,MedGen:C07...</td>\n",
       "      <td>Hyperphenylalaninemia, non-pku;Phenylketonuria...</td>\n",
       "      <td>586</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>102917130</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.12:g.102917130T&gt;C</td>\n",
       "      <td>chr12:g.102917130T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162457</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000546.5(TP53):c.742C&gt;T (p.Arg248Trp)</td>\n",
       "      <td>TP53</td>\n",
       "      <td>HGNC:11998</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Human Phenotype Ontology:HP:0000157,MedGen:C08...</td>\n",
       "      <td>Abnormality of the tongue;Acute myeloid leukem...</td>\n",
       "      <td>12347</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000017.10:g.7577539G&gt;A</td>\n",
       "      <td>chr17:g.7577539G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172160</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_007299.4(BRCA1):c.1018C&gt;T (p.Arg340Ter)</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HGNC:1100</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:CN221562;MedGen:C2676676,OMIM:604370;Me...</td>\n",
       "      <td>Breast and/or ovarian cancer;Breast-ovarian ca...</td>\n",
       "      <td>17675</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>17</td>\n",
       "      <td>41234451</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000017.10:g.41234451G&gt;A</td>\n",
       "      <td>chr17:g.41234451G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242498</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000251.2(MSH2):c.1147C&gt;T (p.Arg383Ter)</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>HGNC:7325</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C0027672,SNOMED CT:699346009;MedGen:C00...</td>\n",
       "      <td>Hereditary cancer-predisposing syndrome;Heredi...</td>\n",
       "      <td>90554</td>\n",
       "      <td>NC_000002.11</td>\n",
       "      <td>2</td>\n",
       "      <td>47656951</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000002.11:g.47656951C&gt;T</td>\n",
       "      <td>chr2:g.47656951C&gt;T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Type                                        Name  \\\n",
       "63376   single nucleotide variant    NM_000314.7(PTEN):c.517C>T (p.Arg173Cys)   \n",
       "141137  single nucleotide variant         NM_000277.3(PAH):c.1A>G (p.Met1Val)   \n",
       "141138  single nucleotide variant         NM_000277.3(PAH):c.1A>G (p.Met1Val)   \n",
       "162457  single nucleotide variant    NM_000546.5(TP53):c.742C>T (p.Arg248Trp)   \n",
       "172160  single nucleotide variant  NM_007299.4(BRCA1):c.1018C>T (p.Arg340Ter)   \n",
       "242498  single nucleotide variant   NM_000251.2(MSH2):c.1147C>T (p.Arg383Ter)   \n",
       "\n",
       "       GeneSymbol     HGNC_ID ClinicalSignificance  \\\n",
       "63376        PTEN   HGNC:9588           Pathogenic   \n",
       "141137        PAH   HGNC:8582           Pathogenic   \n",
       "141138        PAH   HGNC:8582           Pathogenic   \n",
       "162457       TP53  HGNC:11998           Pathogenic   \n",
       "172160      BRCA1   HGNC:1100           Pathogenic   \n",
       "242498       MSH2   HGNC:7325           Pathogenic   \n",
       "\n",
       "                                             PhenotypeIDS  \\\n",
       "63376   MedGen:CN072330,OMIM:158350;MedGen:CN072330,OM...   \n",
       "141137  Human Phenotype Ontology:HP:0004923,MedGen:C07...   \n",
       "141138  Human Phenotype Ontology:HP:0004923,MedGen:C07...   \n",
       "162457  Human Phenotype Ontology:HP:0000157,MedGen:C08...   \n",
       "172160  MedGen:CN221562;MedGen:C2676676,OMIM:604370;Me...   \n",
       "242498  MedGen:C0027672,SNOMED CT:699346009;MedGen:C00...   \n",
       "\n",
       "                                            PhenotypeList  VariationID  \\\n",
       "63376   Cowden syndrome 1;Cowden syndrome 1;Glioma sus...       189500   \n",
       "141137  Hyperphenylalaninemia, non-pku;Phenylketonuria...          586   \n",
       "141138  Hyperphenylalaninemia, non-pku;Phenylketonuria...          586   \n",
       "162457  Abnormality of the tongue;Acute myeloid leukem...        12347   \n",
       "172160  Breast and/or ovarian cancer;Breast-ovarian ca...        17675   \n",
       "242498  Hereditary cancer-predisposing syndrome;Heredi...        90554   \n",
       "\n",
       "       ChromosomeAccession Chromosome      Start ReferenceAllele  \\\n",
       "63376         NC_000010.10         10   89711899               C   \n",
       "141137        NC_000012.11         12  103310908               T   \n",
       "141138        NC_000012.12         12  102917130               T   \n",
       "162457        NC_000017.10         17    7577539               G   \n",
       "172160        NC_000017.10         17   41234451               G   \n",
       "242498        NC_000002.11          2   47656951               C   \n",
       "\n",
       "       AlternateAllele              ReviewStatus Rating  \\\n",
       "63376                T  reviewed by expert panel  three   \n",
       "141137               C  reviewed by expert panel  three   \n",
       "141138               C  reviewed by expert panel  three   \n",
       "162457               A  reviewed by expert panel  three   \n",
       "172160               A  reviewed by expert panel  three   \n",
       "242498               T  reviewed by expert panel  three   \n",
       "\n",
       "                            HGVS_NC              HGVS_chr  \n",
       "63376    NC_000010.10:g.89711899C>T   chr10:g.89711899C>T  \n",
       "141137  NC_000012.11:g.103310908T>C  chr12:g.103310908T>C  \n",
       "141138  NC_000012.12:g.102917130T>C  chr12:g.102917130T>C  \n",
       "162457    NC_000017.10:g.7577539G>A    chr17:g.7577539G>A  \n",
       "172160   NC_000017.10:g.41234451G>A   chr17:g.41234451G>A  \n",
       "242498   NC_000002.11:g.47656951C>T    chr2:g.47656951C>T  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[(df['HGVS_NC'].str.contains(\"NC_000012.11:g.103234177C>T|NC_000002.11:g.47656951C>T|NC_000003.11:g.37038192G>A|NC_000003.11:g.37042536C>T|NC_000003.11:g.37045935C>T|NC_000003.11:g.37048546C>T|NC_000003.11:g.37053589C>T|NC_000003.11:g.37056036G>A|NC_000017.10:g.7577539G>A|NC_000017.10:g.7577548C>T|NC_000017.10:g.7578190T>C|NC_000021.8:g.36171704G>T|NC_000021.8:g.36252962C>G|NC_000021.8:g.36259163T>C|NC_000010.10:g.89711899C>T|NC_000012.11:g.103310908T>C|NC_000012.12:g.102917130T>C|NC_000017.10:g.7577120C>T|NC_000017.10:g.7577538C>T\"))|(df['HGVS_chr'].str.contains(\"chr17:g.41228590G>A|chr17:g.41234451G>A\"))]\n",
    "df.shape # 21 x 17\n",
    "\n",
    "# First 1 NC does not match to Wikidata, chr (2) do not match to Wikidata *18 possible writes of 21\n",
    "df.head(21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7, 17)"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Training data, Feb 11\n",
    "\n",
    "# No match (need item page)\n",
    "## NC_000012.11:g.103234177C>T\n",
    "\n",
    "# Single NC match\n",
    "## NC_000010.10:g.89711899C>T\n",
    "\n",
    "# Single chr match\n",
    "## chr17:g.41234451G>A\n",
    "\n",
    "# Two NC matches in for same Wikidata item page, but different entries\n",
    "## NC_000012.11:g.103310908T>C\n",
    "## NC_000012.12:g.102917130T>C\n",
    "\n",
    "# NC matches to two different Wikidata item pages (remove from https://www.wikidata.org/wiki/Q28371040 later)\n",
    "## NC_000002.11:g.47656951C>T\n",
    "### Removed after testing, but still keeping in training dataset\n",
    "\n",
    "# One NC match, one chr match manually entered (Feb 11)\n",
    "## NC_000017.10:g.7577539G>A\n",
    "## chr17:g.7577539G>A\n",
    "\n",
    "\n",
    "df = df[(df['HGVS_NC'].str.contains(\"NC_000012.11:g.103234177C>T|NC_000010.10:g.89711899C>T|NC_000012.11:g.103310908T>C|NC_000012.12:g.102917130T>C|NC_000002.11:g.47656951C>T|NC_000017.10:g.7577539G>A\"))|(df['HGVS_chr'].str.contains(\"chr17:g.41234451G>A|chr17:g.7577539G>A\"))]\n",
    "df.shape # 7 x 17\n",
    "df.head(7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['Variant Qid'] = \"\" # To be replaced with 'absent' or 'multiple' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'str' object cannot be interpreted as an integer",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-157-7f90fe7d0d01>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# range has to take an inteteger\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"string\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: 'str' object cannot be interpreted as an integer"
     ]
    }
   ],
   "source": [
    "# range has to take an integer (can't be a string, object, etc)\n",
    "\n",
    "for i in range(\"string\"):\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q29938054'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q29938054'}}]\n",
      "[]\n",
      "[]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q64401263'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q64401263'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q64401263'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q64401263'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q28371040'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q28371040'}}\n",
      " {'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q28371040'}}]\n",
      "need item page\n",
      "[]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q29938735'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q28599622'}}]\n",
      "[{'variant': {'type': 'uri', 'value': 'http://www.wikidata.org/entity/Q28599622'}}]\n",
      "The total time of this loop is: 2.8659229278564453 seconds, or 0.04776538213094075 minutes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Rating</th>\n",
       "      <th>HGVS_NC</th>\n",
       "      <th>HGVS_chr</th>\n",
       "      <th>Variant Qid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63376</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000314.7(PTEN):c.517C&gt;T (p.Arg173Cys)</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>HGNC:9588</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:CN072330,OMIM:158350;MedGen:CN072330,OM...</td>\n",
       "      <td>Cowden syndrome 1;Cowden syndrome 1;Glioma sus...</td>\n",
       "      <td>189500</td>\n",
       "      <td>NC_000010.10</td>\n",
       "      <td>10</td>\n",
       "      <td>89711899</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000010.10:g.89711899C&gt;T</td>\n",
       "      <td>chr10:g.89711899C&gt;T</td>\n",
       "      <td>Q29938054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141117</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1315+1G&gt;A</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MeSH:D030342,MedGen:C0950123;MedGen:C0751434,O...</td>\n",
       "      <td>Inborn genetic diseases;Phenylketonuria;not pr...</td>\n",
       "      <td>576</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>103234177</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.11:g.103234177C&gt;T</td>\n",
       "      <td>chr12:g.103234177C&gt;T</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141137</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1A&gt;G (p.Met1Val)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Human Phenotype Ontology:HP:0004923,MedGen:C07...</td>\n",
       "      <td>Hyperphenylalaninemia, non-pku;Phenylketonuria...</td>\n",
       "      <td>586</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>103310908</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.11:g.103310908T&gt;C</td>\n",
       "      <td>chr12:g.103310908T&gt;C</td>\n",
       "      <td>Q64401263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141138</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1A&gt;G (p.Met1Val)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Human Phenotype Ontology:HP:0004923,MedGen:C07...</td>\n",
       "      <td>Hyperphenylalaninemia, non-pku;Phenylketonuria...</td>\n",
       "      <td>586</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>102917130</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.12:g.102917130T&gt;C</td>\n",
       "      <td>chr12:g.102917130T&gt;C</td>\n",
       "      <td>Q64401263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162457</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000546.5(TP53):c.742C&gt;T (p.Arg248Trp)</td>\n",
       "      <td>TP53</td>\n",
       "      <td>HGNC:11998</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>Human Phenotype Ontology:HP:0000157,MedGen:C08...</td>\n",
       "      <td>Abnormality of the tongue;Acute myeloid leukem...</td>\n",
       "      <td>12347</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>17</td>\n",
       "      <td>7577539</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000017.10:g.7577539G&gt;A</td>\n",
       "      <td>chr17:g.7577539G&gt;A</td>\n",
       "      <td>Q28371040</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Type                                      Name  \\\n",
       "63376   single nucleotide variant  NM_000314.7(PTEN):c.517C>T (p.Arg173Cys)   \n",
       "141117  single nucleotide variant              NM_000277.3(PAH):c.1315+1G>A   \n",
       "141137  single nucleotide variant       NM_000277.3(PAH):c.1A>G (p.Met1Val)   \n",
       "141138  single nucleotide variant       NM_000277.3(PAH):c.1A>G (p.Met1Val)   \n",
       "162457  single nucleotide variant  NM_000546.5(TP53):c.742C>T (p.Arg248Trp)   \n",
       "\n",
       "       GeneSymbol     HGNC_ID ClinicalSignificance  \\\n",
       "63376        PTEN   HGNC:9588           Pathogenic   \n",
       "141117        PAH   HGNC:8582           Pathogenic   \n",
       "141137        PAH   HGNC:8582           Pathogenic   \n",
       "141138        PAH   HGNC:8582           Pathogenic   \n",
       "162457       TP53  HGNC:11998           Pathogenic   \n",
       "\n",
       "                                             PhenotypeIDS  \\\n",
       "63376   MedGen:CN072330,OMIM:158350;MedGen:CN072330,OM...   \n",
       "141117  MeSH:D030342,MedGen:C0950123;MedGen:C0751434,O...   \n",
       "141137  Human Phenotype Ontology:HP:0004923,MedGen:C07...   \n",
       "141138  Human Phenotype Ontology:HP:0004923,MedGen:C07...   \n",
       "162457  Human Phenotype Ontology:HP:0000157,MedGen:C08...   \n",
       "\n",
       "                                            PhenotypeList  VariationID  \\\n",
       "63376   Cowden syndrome 1;Cowden syndrome 1;Glioma sus...       189500   \n",
       "141117  Inborn genetic diseases;Phenylketonuria;not pr...          576   \n",
       "141137  Hyperphenylalaninemia, non-pku;Phenylketonuria...          586   \n",
       "141138  Hyperphenylalaninemia, non-pku;Phenylketonuria...          586   \n",
       "162457  Abnormality of the tongue;Acute myeloid leukem...        12347   \n",
       "\n",
       "       ChromosomeAccession Chromosome      Start ReferenceAllele  \\\n",
       "63376         NC_000010.10         10   89711899               C   \n",
       "141117        NC_000012.11         12  103234177               C   \n",
       "141137        NC_000012.11         12  103310908               T   \n",
       "141138        NC_000012.12         12  102917130               T   \n",
       "162457        NC_000017.10         17    7577539               G   \n",
       "\n",
       "       AlternateAllele              ReviewStatus Rating  \\\n",
       "63376                T  reviewed by expert panel  three   \n",
       "141117               T  reviewed by expert panel  three   \n",
       "141137               C  reviewed by expert panel  three   \n",
       "141138               C  reviewed by expert panel  three   \n",
       "162457               A  reviewed by expert panel  three   \n",
       "\n",
       "                            HGVS_NC              HGVS_chr Variant Qid  \n",
       "63376    NC_000010.10:g.89711899C>T   chr10:g.89711899C>T   Q29938054  \n",
       "141117  NC_000012.11:g.103234177C>T  chr12:g.103234177C>T              \n",
       "141137  NC_000012.11:g.103310908T>C  chr12:g.103310908T>C   Q64401263  \n",
       "141138  NC_000012.12:g.102917130T>C  chr12:g.102917130T>C   Q64401263  \n",
       "162457    NC_000017.10:g.7577539G>A    chr17:g.7577539G>A   Q28371040  "
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For loop on training dataset\n",
    "start_time = time.time() # Keep track of how long it takes loop to run\n",
    "\n",
    "for index, row in df.iterrows(): # Index is a row number, row is all variables and values for that row\n",
    "    \n",
    "    # Identify string for a given data table row to retrive the variant name and item page Qid from Wikidata\n",
    "    HGVS_NC = df.loc[index, 'HGVS_NC']\n",
    "    HGVS_chr = df.loc[index, 'HGVS_chr']\n",
    "    \n",
    "    # SparQL query to search HGVS Identifier (P3331) in Wikidata \n",
    "    sparqlQuery_HGVSNC = \"SELECT * WHERE {?variant wdt:P3331 \\\"\"+HGVS_NC+\"\\\"}\" \n",
    "    result_HGVSNC = wdi_core.WDItemEngine.execute_sparql_query(sparqlQuery_HGVSNC) \n",
    "    sparqlQuery_HGVSchr = \"SELECT * WHERE {?variant wdt:P3331 \\\"\"+HGVS_chr+\"\\\"}\" \n",
    "    result_HGVSchr = wdi_core.WDItemEngine.execute_sparql_query(sparqlQuery_HGVSchr)   \n",
    "    \n",
    "    # Assign the results to respective lists *we know these are lists by using the type() function\n",
    "    NC = result_HGVSNC[\"results\"][\"bindings\"]\n",
    "    CHR = result_HGVSchr[\"results\"][\"bindings\"]\n",
    "    \n",
    "    # Concatenate lists as a combined array based on being different (2+), empty (0), or one or duplicate (1)\n",
    "    HGVSarray = np.concatenate((NC,CHR), axis=0)\n",
    "    print(HGVSarray)\n",
    "    \n",
    "    HGVSlist = [] # create empty list\n",
    "    # Successful execution of code doesn't mean it's working as it should.\n",
    "    \n",
    "    for i in range(len(HGVSarray)): # assigning each list an integer\n",
    "        print(HGVSlist)\n",
    "        if HGVSarray[i] not in HGVSarray[i + 1:]: # if 0 is not in 1-[however long it is], and goes through all\n",
    "            HGVSlist.append(HGVSarray[i]) # then add to the list (shows whatever is unique)\n",
    "            \n",
    "    print(HGVSlist)\n",
    "            \n",
    "            \n",
    "            # Scenario 1: Flag if HGVS identifier(s) have multiple wikidata item pages (works, tested NC_000002.11:g.47656951C>T)\n",
    "    if len(HGVSlist) > 1: \n",
    "        print(\"error\") \n",
    "        continue\n",
    "                \n",
    "            # Scenario 2: Go forward with write if only one or if duplicate (combine)     \n",
    "    Variant_Qid = HGVSlist[0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")\n",
    "    df.at[index, 'Variant Qid'] = Variant_Qid  \n",
    "            \n",
    "            # Add missing NC or chr (HGVS nomenclature: P3331)\n",
    "            \n",
    "            # Add VariationID (ClinVar Variation ID: P3329)\n",
    "            \n",
    "        # Scenario 3: Write item page if no identifiers available   \n",
    "        ## How to create this?\n",
    "        if len(HGVSlist) == 0:\n",
    "            print(\"need item page\")\n",
    "            df.at[index, 'Variant Qid'] = \"absent\"  \n",
    "            #try_write(HGVS_NC, \n",
    "            #          edit_summary=(\"gene variant in human gene \", df.loc[index, 'GeneSymbol']),\n",
    "            #          record_id = Qid,\n",
    "            #          record_prop = 'human gene variant',\n",
    "            #          login = login)\n",
    "            #print(\"item page created\")\n",
    "            \n",
    "        \n",
    "        \n",
    "end_time = time.time() # Captures when loop run ends\n",
    "print(\"The total time of this loop is:\", end_time - start_time, \"seconds, or\", (end_time - start_time)/60, \"minutes\")\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Item page: Q29938054 , for NC_000010.10:g.89711899C>T\n",
      "error: need item page\n",
      "Item page: Q64401263 , for NC_000012.11:g.103310908T>C\n",
      "Item page: Q64401263 , for NC_000012.12:g.102917130T>C\n",
      "same\n",
      "Item page: Q28371040 , for NC_000017.10:g.7577539G>A\n",
      "Item page: Q28371040 , for chr17:g.7577539G>A\n",
      "Item page: Q29938735 , for chr17:g.41234451G>A\n",
      "Item page: Q28599622 , for NC_000002.11:g.47656951C>T\n",
      "The total time of this loop is: 3.8079538345336914 seconds, or 0.06346589724222819 minutes\n"
     ]
    }
   ],
   "source": [
    "# For loop on training dataset\n",
    "start_time = time.time() # Keep track of how long it takes loop to run\n",
    "\n",
    "\n",
    "for index, row in df.iterrows(): # Index is a row number, row is all variables and values for that row\n",
    "    \n",
    "    \n",
    "    # Identify string for a given row to retrive variant name and item page Qid from Wikidata\n",
    "    HGVS_NC = df.loc[index, 'HGVS_NC']\n",
    "    HGVS_chr = df.loc[index, 'HGVS_chr']\n",
    "    \n",
    "    # SparQL query to search HGVS Identifier in Wikidata based on P3331 \n",
    "    sparqlQuery_HGVSNC = \"SELECT * WHERE {?variant wdt:P3331 \\\"\"+HGVS_NC+\"\\\"}\" \n",
    "    result_HGVSNC = wdi_core.WDItemEngine.execute_sparql_query(sparqlQuery_HGVSNC) \n",
    "    sparqlQuery_HGVSchr = \"SELECT * WHERE {?variant wdt:P3331 \\\"\"+HGVS_chr+\"\\\"}\" \n",
    "    result_HGVSchr = wdi_core.WDItemEngine.execute_sparql_query(sparqlQuery_HGVSchr)   \n",
    "    \n",
    "    # Assign a length to the resultant dictionary for either NC or chr (number of item pages by Qid)\n",
    "    NC_qlength = len(result_HGVSNC[\"results\"][\"bindings\"]) \n",
    "    chr_qlength = len(result_HGVSchr[\"results\"][\"bindings\"]) \n",
    "    # Create a unique id based on the joined length\n",
    "    combo_qlength = NC_qlength + chr_qlength # concatenate the values and then get the unique values over the number\n",
    "    # create an array with both together over having integers\n",
    "        # then 2, 1, or 0 situation... but may be tricky to assign a set here \n",
    "        # an array of objects\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    # If there is more than one Wikidata item page that matches with the HGVS identifiers\n",
    "    ## Assumption: HGVS identifier is unique to that variant\n",
    "    if combo_qlength > 2:\n",
    "        print(\"error: multiple wikidata item pages for HGVS identifier\") # log in output\n",
    "        continue\n",
    "    \n",
    "    # If there is no Wikidata item page that matches with the HGVS identifiers\n",
    "    ## How to create the item page, if there isn't a variant name automatically in data table?\n",
    "        # Use the NC name, and people will update later\n",
    "        # Need name (NC)\n",
    "        # Description: \"gene variant in human gene X\"\n",
    "    ## How to best re-work through loop after adding HGVS item page?\n",
    "        # Only go through the loop once for every line\n",
    "        \n",
    "    if combo_qlength == 0:\n",
    "        print(\"error: need item page\") # Assume no item page exists\n",
    "        \n",
    "        \n",
    "        \n",
    "    # If there \n",
    "    ## How to make this all less repetitive? (adding identifier and clinvar id)\n",
    "    if combo_qlength > 0 & combo_qlength < 3:\n",
    "        if NC_qlength == chr_qlength: \n",
    "        ## How to best address if they're not equal?\n",
    "            \n",
    "            NC_qid = result_HGVSNC[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")\n",
    "            chr_qid = result_HGVSchr[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")    \n",
    "            print(\"same\")\n",
    "        if NC_qlength == 1:\n",
    "        ## Note there are two NCs for Q64401263\n",
    "            # Non-write\n",
    "            # Additional disease, if applicable\n",
    "            NC_qid = result_HGVSNC[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")\n",
    "            print(\"Item page:\", NC_qid, \", for\", HGVS_NC) \n",
    "            # Add chr identifier\n",
    "        if chr_qlength == 1:\n",
    "            chr_qid = result_HGVSchr[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")    \n",
    "            print(\"Item page:\", chr_qid, \", for\", HGVS_chr)\n",
    "            ## Do this later\n",
    "            # Add nc identifier\n",
    "            #identifier_NC = [wdi_core.WDItemID(value=chr_qid, prop_nr=\"P3331\")] \n",
    "            #wikidata_NCidentifer = wdi_core.WDItemEngine(wd_item_id=chr_qid, \n",
    "             #                                     data=identifier_NC, \n",
    "              #                                    append_value=[\"P3331\"])  \n",
    "            #wikidata_NCidentifer.get_wd_json_representation() \n",
    "            #wikidata_NCidentifer.write(login) \n",
    "        ## Edits to reference\n",
    "\n",
    "        ## Add VariationID - example, but no reference url in data... leave out? https://www.wikidata.org/wiki/Q21851559\n",
    "        \n",
    "\n",
    "        \n",
    "        \n",
    "# look at diseases (add genetic association)\n",
    "# Symmetry for diseases\n",
    "        \n",
    "# Separate out Phenotype List (;) for disease \n",
    "    # Split one line into two (ie if there are two diseases)\n",
    "        # Split by commas for identifiers\n",
    "            # For example, if there are four identifiers\n",
    "            # Take all identifiers for that disease, and search against Wikidata\n",
    "            # Take union of results\n",
    "                # If only one QiD found, then good\n",
    "                # If there are multiple, then flag (don't write anything) *dont write anything for whole line\n",
    "                # Add identifiers from dataset that are missing in Wikidata\n",
    "                \n",
    "\n",
    "# array of variances, with one variant shown here   \n",
    "# concatenate to get this below (with nc and chr)\n",
    "# then write something that removes duplicates: 0, 1, or 2\n",
    "    \n",
    "    \n",
    "        \n",
    "end_time = time.time() # Captures when loop run ends\n",
    "print(\"The total time of this loop is:\", end_time - start_time, \"seconds, or\", (end_time - start_time)/60, \"minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_000002.11:g.47656951C>T',\n",
       " 'NC_000003.11:g.37038192G>A',\n",
       " 'NC_000003.11:g.37042536C>T',\n",
       " 'NC_000003.11:g.37045935C>T',\n",
       " 'NC_000003.11:g.37048546C>T',\n",
       " 'NC_000003.11:g.37053589C>T',\n",
       " 'NC_000003.11:g.37056036G>A',\n",
       " 'NC_000010.10:g.89711899C>T',\n",
       " 'NC_000012.11:g.103310908T>C',\n",
       " 'NC_000012.12:g.102917130T>C',\n",
       " 'NC_000017.10:g.7577120C>T',\n",
       " 'NC_000017.10:g.7577538C>T',\n",
       " 'NC_000017.10:g.7577539G>A',\n",
       " 'NC_000017.10:g.7577548C>T',\n",
       " 'NC_000017.10:g.7578190T>C',\n",
       " 'NC_000021.8:g.36171704G>T',\n",
       " 'NC_000021.8:g.36252962C>G',\n",
       " 'NC_000021.8:g.36259163T>C'}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Download HGVS Wikidata query results as query.csv \n",
    "### https://query.wikidata.org/#SELECT%20%2a%20WHERE%20%7B%3Fgene%20wdt%3AP3331%20%3FHGVS%7D\n",
    "\n",
    "query = pd.read_csv(\"/Users/sulhasan/Desktop/Su Lab Projects/ClinVar-Bot_GeneWikiCentral-Issue50/query.csv\")  \n",
    "\n",
    "## Subset for NC or chr\n",
    "query = query[query['HGVS'].str.contains('chr|NC_',  na=False)] # 818 with NC or chr\n",
    "query.head(5) \n",
    "\n",
    "# How many HGVS IDs match for Wikidata query (818) vs. manually created in ClinVar (3778 x 2 for NC or chr)\n",
    "len(set(query[\"HGVS\"]) & (set(df[\"HGVS_chr\"]) | set(df[\"HGVS_NC\"]))) # 17\n",
    "set(query[\"HGVS\"]) & (set(df[\"HGVS_chr\"]) | set(hasHGVS[\"HGVS_NC\"])) # 4 chr, 13 NC *chr dont match..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep anything with 'two' or more stars in 'Rating' column\n",
    "twoplus = new[~new['Rating'].str.contains('one')] # excludes both one and none\n",
    "twoplus.shape # 13 columns, 196530 rows: 1123285 removed, 14.9 % of all data usable\n",
    "# Keep anything noted as 'single nucleotide variant' in 'Type' column\n",
    "snv = twoplus[twoplus['Type'].str.contains('single nucleotide variant')]\n",
    "snv.shape # 174733 rows, or 88.9% of data with two or more stars (13.2% of all data)\n",
    "## 83.2 % of total data, prior to star rating filter, are SNVs\n",
    "# Keep anything with 'Pathogenic' in the 'ClinicalSignificance' column\n",
    "patho = snv[snv['ClinicalSignificance'].str.contains('Pathogenic')]\n",
    "patho.shape # 20040 rows, or 10.2% of data with two or more stars that are snvs (1.5% of all data)\n",
    "## 12.8 % of total data, prior to star rating filter, are Pathogenic"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
