{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scheduled Integration of ClinVar Gene Variant-Disease Data into WikiData\n",
    "\n",
    "ClinVar aggregates information about genomic variation and its relationship to human health <br>\n",
    "CC0 https://www.ncbi.nlm.nih.gov/clinvar/\n",
    "\n",
    "This scheduled bot operates monthly through WDI to integrate ClinVar Gene Variant-Disease Data <br>\n",
    "https://www.ncbi.nlm.nih.gov/clinvar/docs/ftp_primer/ (variant_summary) <br>\n",
    "https://github.com/SuLab/GeneWikiCentral/issues/50 <br>\n",
    "http://jenkins.sulab.org/ <br>\n",
    "\n",
    "Python script contributions, in order: Sabah Ul-Hasan, Andrew I Su, Tong Shu Li"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Checks and Tests\n",
    "\n",
    "- x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relevant Modules and Libraries\n",
    "\n",
    "import os # OS package to ensure interaction between the modules (ie WDI) and current OS being used\n",
    "\n",
    "from datetime import datetime # For identifying the current date and time\n",
    "import time # Keep track of total for loop run time\n",
    "\n",
    "import gzip # For unzip of files\n",
    "import shutil # Copies content of source file(s)\n",
    "import csv # For converting file(s) to csv format\n",
    "\n",
    "import pandas as pd # For data organization, abbreviated to pd\n",
    "\n",
    "from wikidataintegrator import wdi_core, wdi_login # Core and login from wikidataintegrator module\n",
    "from wikidataintegrator.ref_handlers import update_retrieved_if_new_multiple_refs # For retrieving references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM Allelic Variant:613653.0001</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM Allelic Variant:613653.0001</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM Allelic Variant:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM Allelic Variant:613653.0002</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>150829393</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>15q25</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM Allelic Variant:613624.0001,UniProtKB (pr...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15041                      indel   \n",
       "1      15041                      indel   \n",
       "2      15042                   deletion   \n",
       "3      15042                   deletion   \n",
       "4      15043  single nucleotide variant   \n",
       "\n",
       "                                                Name  GeneID GeneSymbol  \\\n",
       "0  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "1  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "2     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "3     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "4       NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "\n",
       "      HGNC_ID    ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
       "0  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "1  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "2  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "3  HGNC:22197              Pathogenic              1  Jun 29, 2010   \n",
       "4  HGNC:28986  Uncertain significance              0  Jun 29, 2015   \n",
       "\n",
       "   RS# (dbSNP)  ...  ReferenceAllele         AlternateAllele Cytogenetic  \\\n",
       "0    397704705  ...             GGAT  TGCTGTAAACTGTAACTGTAAA      7p22.1   \n",
       "1    397704705  ...             GGAT  TGCTGTAAACTGTAACTGTAAA      7p22.1   \n",
       "2    397704709  ...  GCTGCTGGACCTGCC                       G      7p22.1   \n",
       "3    397704709  ...  GCTGCTGGACCTGCC                       G      7p22.1   \n",
       "4    150829393  ...                G                       A       15q25   \n",
       "\n",
       "                     ReviewStatus NumberSubmitters Guidelines TestedInGTR  \\\n",
       "0  no assertion criteria provided                1        NaN           N   \n",
       "1  no assertion criteria provided                1        NaN           N   \n",
       "2  no assertion criteria provided                1        NaN           N   \n",
       "3  no assertion criteria provided                1        NaN           N   \n",
       "4  no assertion criteria provided                1        NaN           N   \n",
       "\n",
       "                                            OtherIDs SubmitterCategories  \\\n",
       "0                   OMIM Allelic Variant:613653.0001                   1   \n",
       "1                   OMIM Allelic Variant:613653.0001                   1   \n",
       "2                   OMIM Allelic Variant:613653.0002                   1   \n",
       "3                   OMIM Allelic Variant:613653.0002                   1   \n",
       "4  OMIM Allelic Variant:613624.0001,UniProtKB (pr...                   1   \n",
       "\n",
       "   VariationID  \n",
       "0            2  \n",
       "1            2  \n",
       "2            3  \n",
       "3            3  \n",
       "4            4  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download data from NCBI\n",
    "\n",
    "## Make sure os has wget installed, or the following command wont work\n",
    "os.system('wget ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz') \n",
    "timeStringNow = datetime.now().strftime(\"+%Y-%m-%dT00:00:00Z\") # time stamp of download \n",
    "\n",
    "## Unzip the file\n",
    "with gzip.open('variant_summary.txt.gz', 'rb') as f_in:\n",
    "    with open('variant_summary.txt', 'wb') as f_out:\n",
    "        shutil.copyfileobj(f_in, f_out)\n",
    "        \n",
    "## Convert .txt file to .csv\n",
    "txt_file = r\"variant_summary.txt\"\n",
    "csv_file = r\"variant_summary.csv\"\n",
    "\n",
    "with open(txt_file, \"r\") as in_text:\n",
    "    in_reader = csv.reader(in_text, delimiter = '\\t')\n",
    "    with open(csv_file, \"w\") as out_csv:\n",
    "        out_writer = csv.writer(out_csv)\n",
    "        for row in in_reader:\n",
    "            out_writer.writerow(row)\n",
    "\n",
    "## Import .csv file and read first 5 rows\n",
    "df = pd.read_csv(\"variant_summary.csv\") \n",
    "df.shape # 31 columns, 1330018 rows\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  # Remove the CWD from sys.path while we load stuff.\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/Users/sulhasan/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Rating</th>\n",
       "      <th>HGVS_NC</th>\n",
       "      <th>HGVS_chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1088</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1315+1G&gt;A</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MeSH:D030342,MedGen:C0950123;MedGen:C0751434,O...</td>\n",
       "      <td>Inborn genetic diseases;Phenylketonuria;not pr...</td>\n",
       "      <td>576</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>103234177</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.11:g.103234177C&gt;T</td>\n",
       "      <td>chr12:g.103234177C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1089</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1315+1G&gt;A</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MeSH:D030342,MedGen:C0950123;MedGen:C0751434,O...</td>\n",
       "      <td>Inborn genetic diseases;Phenylketonuria;not pr...</td>\n",
       "      <td>576</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>102840399</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.12:g.102840399C&gt;T</td>\n",
       "      <td>chr12:g.102840399C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1222C&gt;T (p.Arg408Trp)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C0751434,OMIM:261600,Orphanet:ORPHA716,...</td>\n",
       "      <td>Phenylketonuria;not provided</td>\n",
       "      <td>577</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>103234271</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.11:g.103234271G&gt;A</td>\n",
       "      <td>chr12:g.103234271G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1091</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.1222C&gt;T (p.Arg408Trp)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C0751434,OMIM:261600,Orphanet:ORPHA716,...</td>\n",
       "      <td>Phenylketonuria;not provided</td>\n",
       "      <td>577</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>102840493</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.12:g.102840493G&gt;A</td>\n",
       "      <td>chr12:g.102840493G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1098</th>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000277.3(PAH):c.331C&gt;T (p.Arg111Ter)</td>\n",
       "      <td>PAH</td>\n",
       "      <td>HGNC:8582</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>MedGen:C0751434,OMIM:261600,Orphanet:ORPHA716,...</td>\n",
       "      <td>Phenylketonuria;not provided</td>\n",
       "      <td>581</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>103288534</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>reviewed by expert panel</td>\n",
       "      <td>three</td>\n",
       "      <td>NC_000012.11:g.103288534G&gt;A</td>\n",
       "      <td>chr12:g.103288534G&gt;A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Type                                      Name  \\\n",
       "1088  single nucleotide variant              NM_000277.3(PAH):c.1315+1G>A   \n",
       "1089  single nucleotide variant              NM_000277.3(PAH):c.1315+1G>A   \n",
       "1090  single nucleotide variant  NM_000277.3(PAH):c.1222C>T (p.Arg408Trp)   \n",
       "1091  single nucleotide variant  NM_000277.3(PAH):c.1222C>T (p.Arg408Trp)   \n",
       "1098  single nucleotide variant   NM_000277.3(PAH):c.331C>T (p.Arg111Ter)   \n",
       "\n",
       "     GeneSymbol    HGNC_ID ClinicalSignificance  \\\n",
       "1088        PAH  HGNC:8582           Pathogenic   \n",
       "1089        PAH  HGNC:8582           Pathogenic   \n",
       "1090        PAH  HGNC:8582           Pathogenic   \n",
       "1091        PAH  HGNC:8582           Pathogenic   \n",
       "1098        PAH  HGNC:8582           Pathogenic   \n",
       "\n",
       "                                           PhenotypeIDS  \\\n",
       "1088  MeSH:D030342,MedGen:C0950123;MedGen:C0751434,O...   \n",
       "1089  MeSH:D030342,MedGen:C0950123;MedGen:C0751434,O...   \n",
       "1090  MedGen:C0751434,OMIM:261600,Orphanet:ORPHA716,...   \n",
       "1091  MedGen:C0751434,OMIM:261600,Orphanet:ORPHA716,...   \n",
       "1098  MedGen:C0751434,OMIM:261600,Orphanet:ORPHA716,...   \n",
       "\n",
       "                                          PhenotypeList  VariationID  \\\n",
       "1088  Inborn genetic diseases;Phenylketonuria;not pr...          576   \n",
       "1089  Inborn genetic diseases;Phenylketonuria;not pr...          576   \n",
       "1090                       Phenylketonuria;not provided          577   \n",
       "1091                       Phenylketonuria;not provided          577   \n",
       "1098                       Phenylketonuria;not provided          581   \n",
       "\n",
       "     ChromosomeAccession Chromosome      Start ReferenceAllele  \\\n",
       "1088        NC_000012.11         12  103234177               C   \n",
       "1089        NC_000012.12         12  102840399               C   \n",
       "1090        NC_000012.11         12  103234271               G   \n",
       "1091        NC_000012.12         12  102840493               G   \n",
       "1098        NC_000012.11         12  103288534               G   \n",
       "\n",
       "     AlternateAllele              ReviewStatus Rating  \\\n",
       "1088               T  reviewed by expert panel  three   \n",
       "1089               T  reviewed by expert panel  three   \n",
       "1090               A  reviewed by expert panel  three   \n",
       "1091               A  reviewed by expert panel  three   \n",
       "1098               A  reviewed by expert panel  three   \n",
       "\n",
       "                          HGVS_NC              HGVS_chr  \n",
       "1088  NC_000012.11:g.103234177C>T  chr12:g.103234177C>T  \n",
       "1089  NC_000012.12:g.102840399C>T  chr12:g.102840399C>T  \n",
       "1090  NC_000012.11:g.103234271G>A  chr12:g.103234271G>A  \n",
       "1091  NC_000012.12:g.102840493G>A  chr12:g.102840493G>A  \n",
       "1098  NC_000012.11:g.103288534G>A  chr12:g.103288534G>A  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean-up of data for identification and integration\n",
    "\n",
    "## Columns to keep\n",
    "new = df[['Type', 'Name', 'GeneSymbol','HGNC_ID', 'ClinicalSignificance', 'PhenotypeIDS', 'PhenotypeList','VariationID', \n",
    "            'ChromosomeAccession', 'Chromosome', 'Start', 'ReferenceAllele', 'AlternateAllele',\n",
    "            'ReviewStatus']]\n",
    "\n",
    "## Create new column that converts 'ReviewStatus' to star rating\n",
    "### https://www.ncbi.nlm.nih.gov/clinvar/docs/review_status/\n",
    "new['Rating'] = \"\" # Create empty column for gold star rating\n",
    "## Convert strings from 'ReviewStatus' to 'Rating' \n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion provided'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion criteria provided'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('no assertion for the individual variant'), 'Rating'] = 'none'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, single submitter'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, conflicting interpretations'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, conflicting interpretations'), 'Rating'] = 'one'\n",
    "new.loc[new['ReviewStatus'].str.contains('criteria provided, multiple submitters, no conflicts'), 'Rating'] = 'two'\n",
    "new.loc[new['ReviewStatus'].str.contains('reviewed by expert panel'), 'Rating'] = 'three'\n",
    "new.loc[new['ReviewStatus'].str.contains('practice guideline'), 'Rating'] = 'four'\n",
    "\n",
    "new.shape # 15 columns, 1330018 rows \n",
    "\n",
    "## Training dataset, based on the following criteria\n",
    "threeplus=new[new['Rating'].str.contains('three|four')]\n",
    "snv=threeplus[threeplus['Type'].str.contains('single nucleotide variant')]\n",
    "patho=snv[snv['ClinicalSignificance'].str.contains('Pathogenic')]\n",
    "\n",
    "patho.shape # 15 columns, 3870 rows (0.29% of original dataset)\n",
    "\n",
    "## Create HGVS columns and IDs (based on genomic position)\n",
    "### Naming nomenclature: https://varnomen.hgvs.org/bg-material/numbering/\n",
    "patho['HGVS_NC'] = \"\" # Create empty column for HGVS nomenclature with NC\n",
    "patho['HGVS_chr'] = \"\" # Create empty column for HGVS nomenclature with chr\n",
    "\n",
    "patho['HGVS_NC']=patho['ChromosomeAccession']+':g.'+patho['Start'].astype(str)+patho['ReferenceAllele']+'>'+patho['AlternateAllele']\n",
    "patho['HGVS_chr']='chr'+patho['Chromosome'].astype(str)+':g.'+patho['Start'].astype(str)+patho['ReferenceAllele']+'>'+patho['AlternateAllele']\n",
    "\n",
    "df=patho\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logging in...\n",
      "https://www.wikidata.org/w/api.php\n",
      "Successfully logged in as Sulhasan\n"
     ]
    }
   ],
   "source": [
    "# Login for running WDI\n",
    "print(\"Logging in...\") \n",
    "\n",
    "# **remove lines when scheduling to Jenkins** Enter your own username and password \n",
    "os.environ[\"WDUSER\"] = \"username\" # Uses os package to call and set the environment for wikidata username\n",
    "os.environ[\"WDPASS\"] = \"password\"\n",
    "\n",
    "## Conditional that outputs error command if not in the local python environment\n",
    "if \"WDUSER\" in os.environ and \"WDPASS\" in os.environ: \n",
    "    WDUSER = os.environ['WDUSER']\n",
    "    WDPASS = os.environ['WDPASS']\n",
    "else: \n",
    "    raise ValueError(\"WDUSER and WDPASS must be specified in local.py or as environment variables\")      \n",
    "\n",
    "## Sets attributed username and password as 'login'\n",
    "login = wdi_login.WDLogin(WDUSER, WDPASS) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19, 17)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[(df['HGVS_NC'].str.contains(\"NC_000012.11:g.103234177C>T|NC_000002.11:g.47656951C>T|NC_000003.11:g.37038192G>A|NC_000003.11:g.37042536C>T|NC_000003.11:g.37045935C>T|NC_000003.11:g.37048546C>T|NC_000003.11:g.37053589C>T|NC_000003.11:g.37056036G>A|NC_000017.10:g.7577539G>A|NC_000017.10:g.7577548C>T|NC_000017.10:g.7578190T>C|NC_000021.8:g.36171704G>T|NC_000021.8:g.36252962C>G|NC_000021.8:g.36259163T>C|NC_000010.10:g.89711899C>T|NC_000012.11:g.103310908T>C|NC_000012.12:g.102917130T>C|NC_000017.10:g.7577120C>T|NC_000017.10:g.7577538C>T\"))|(df['HGVS_chr'].str.contains(\"chr17:g.41228590G>A|chr17:g.41234451G>A\"))]\n",
    "df.shape\n",
    "\n",
    "# First 1 NC does not match to Wikidata, chr (2) do not match to Wikidata *18 possible writes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Need item page\n",
      "NC Qid Q64401263\n",
      "Need chr in HGVS\n",
      "NC Qid Q64401263\n",
      "Need chr in HGVS\n",
      "NC Qid Q28371040\n",
      "Need chr in HGVS\n",
      "NC Qid Q28371039\n",
      "Need chr in HGVS\n",
      "NC Qid Q29938325\n",
      "Need chr in HGVS\n",
      "NC Qid Q28371044\n",
      "Need chr in HGVS\n",
      "NC Qid Q28371675\n",
      "Need chr in HGVS\n",
      "NC Qid Q28371676\n",
      "Need chr in HGVS\n",
      "NC Qid Q28371680\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599642\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599631\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599630\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599645\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599641\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599625\n",
      "Need chr in HGVS\n",
      "NC Qid Q28599622\n",
      "Need chr in HGVS\n",
      "NC Qid Q29938384\n",
      "Need chr in HGVS\n",
      "NC Qid Q29938054\n",
      "Need chr in HGVS\n",
      "The total time of this loop is: 9.418941974639893 seconds, or 0.15698236624399822 minutes\n"
     ]
    }
   ],
   "source": [
    "# For loop on training dataset \n",
    "start_time = time.time() # Keep track of how long it takes loop to run\n",
    "\n",
    "for index, row in df.iterrows(): # Index is a row number, row is all variables and values for that row\n",
    "    \n",
    "    # Identify string for given row to retrive variant name and item page Qid from Wikidata\n",
    "    HGVS_NC = df.loc[index, 'HGVS_NC']\n",
    "    HGVS_chr = df.loc[index, 'HGVS_chr']\n",
    "    \n",
    "    # SparQL query to search HGVS Identifier in Wikidata based on P3331 \n",
    "    sparqlQuery_HGVSNC = \"SELECT * WHERE {?variant wdt:P3331 \\\"\"+HGVS_NC+\"\\\"}\" \n",
    "    result_HGVSNC = wdi_core.WDItemEngine.execute_sparql_query(sparqlQuery_HGVSNC) \n",
    "    sparqlQuery_HGVSchr = \"SELECT * WHERE {?variant wdt:P3331 \\\"\"+HGVS_chr+\"\\\"}\" \n",
    "    result_HGVSchr = wdi_core.WDItemEngine.execute_sparql_query(sparqlQuery_HGVSchr)   \n",
    "    \n",
    "    # Assign a length to the resultant dictionary for either NC or chr (number of Qid)\n",
    "    NC_qlength = len(result_HGVSNC[\"results\"][\"bindings\"]) \n",
    "    chr_qlength = len(result_HGVSchr[\"results\"][\"bindings\"]) \n",
    "    \n",
    "    # Match HGVS NC* or chr* created for ClinVar dataset to Wikidata Qid item page\n",
    "    if NC_qlength == 0 & chr_qlength == 0:\n",
    "        print(\"Need item page\") # Write item page here, log (assumption no item page exists) *add HGVS\n",
    "        continue\n",
    "        \n",
    "    if NC_qlength > 1:\n",
    "        print(\"error: multiple qid for nc\") # log\n",
    "        if chr_qlength > 1:\n",
    "            print(\"error: multiple qid for chr\") # log\n",
    "            continue\n",
    "    \n",
    "    if chr_qlength > 1:\n",
    "        print(\"error: multiple qid for chr\") # log\n",
    "        continue\n",
    "        \n",
    "    if NC_qlength == 1:\n",
    "        NC_qid = result_HGVSNC[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")\n",
    "        print(\"NC Qid\", NC_qid)\n",
    "        # Add ClinVar Variation ID (stated in, ref, retrieved)\n",
    "        if chr_qlength == 0:\n",
    "            print(\"Need chr in HGVS\") # Write identifier, keep working through loop\n",
    "        if chr_qlength == 1:\n",
    "            chr_qid = result_HGVSchr[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")\n",
    "            if NC_qid == chr_qid:\n",
    "                print(\"same qid\") # Keep working through loop\n",
    "                # look at diseases (add genetic association)\n",
    "                # Symmetry for diseases\n",
    "                \n",
    "# Separate out Phenotype List (;) for disease \n",
    "    # Split one line into two (ie if there are two diseases)\n",
    "        # Split by commas for identifiers\n",
    "            # For example, if there are four identifiers\n",
    "            # Take all identifiers for that disease, and search against Wikidata\n",
    "            # Take union of results\n",
    "                # If only one QiD found, then good\n",
    "                # If there are multiple, then flag (don't write anything) *dont write anything for whole line\n",
    "                # Add identifiers from dataset that are missing in Wikidata\n",
    "                \n",
    "            else:\n",
    "                print(\"error: different qid\")\n",
    "                continue\n",
    "    \n",
    "    if chr_qlength == 1:\n",
    "        chr_qid = result_HGVSchr[\"results\"][\"bindings\"][0][\"variant\"][\"value\"].replace(\"http://www.wikidata.org/entity/\", \"\")\n",
    "        print(\"chr Qid\", chr_qid)\n",
    "        print(\"Need NC in HGVS\") # Add genetic association, assumption NC_qlength is 0 based on above\n",
    "        # Add ClinVar Variation ID (stated in, ref, retrieved)\n",
    "        \n",
    "# NC, chr (HGVS identifiers)\n",
    "# 0, 0 = absent => write item page\n",
    "# 1, 0 = result => continue\n",
    "# 0, 1 = result => continue\n",
    "# 1, 1 = same => continue\n",
    "# 1, 1 = different => error\n",
    "        \n",
    "end_time = time.time() # Captures when loop run ends\n",
    "print(\"The total time of this loop is:\", end_time - start_time, \"seconds, or\", (end_time - start_time)/60, \"minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NC_000002.11:g.47656951C>T',\n",
       " 'NC_000003.11:g.37038192G>A',\n",
       " 'NC_000003.11:g.37042536C>T',\n",
       " 'NC_000003.11:g.37045935C>T',\n",
       " 'NC_000003.11:g.37048546C>T',\n",
       " 'NC_000003.11:g.37053589C>T',\n",
       " 'NC_000003.11:g.37056036G>A',\n",
       " 'NC_000010.10:g.89711899C>T',\n",
       " 'NC_000012.11:g.103310908T>C',\n",
       " 'NC_000012.12:g.102917130T>C',\n",
       " 'NC_000017.10:g.7577120C>T',\n",
       " 'NC_000017.10:g.7577538C>T',\n",
       " 'NC_000017.10:g.7577539G>A',\n",
       " 'NC_000017.10:g.7577548C>T',\n",
       " 'NC_000017.10:g.7578190T>C',\n",
       " 'NC_000021.8:g.36171704G>T',\n",
       " 'NC_000021.8:g.36252962C>G',\n",
       " 'NC_000021.8:g.36259163T>C'}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Download HGVS Wikidata query results as query.csv \n",
    "### https://query.wikidata.org/#SELECT%20%2a%20WHERE%20%7B%3Fgene%20wdt%3AP3331%20%3FHGVS%7D\n",
    "\n",
    "query = pd.read_csv(\"/Users/sulhasan/Desktop/Su Lab Projects/ClinVar-Bot_GeneWikiCentral-Issue50/query.csv\")  \n",
    "\n",
    "## Subset for NC or chr\n",
    "query = query[query['HGVS'].str.contains('chr|NC_',  na=False)] # 818 with NC or chr\n",
    "query.head(5) \n",
    "\n",
    "# How many HGVS IDs match for Wikidata query (818) vs. manually created in ClinVar (3778 x 2 for NC or chr)\n",
    "len(set(query[\"HGVS\"]) & (set(df[\"HGVS_chr\"]) | set(df[\"HGVS_NC\"]))) # 17\n",
    "set(query[\"HGVS\"]) & (set(df[\"HGVS_chr\"]) | set(hasHGVS[\"HGVS_NC\"])) # 4 chr, 13 NC *chr dont match..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep anything with 'two' or more stars in 'Rating' column\n",
    "twoplus = new[~new['Rating'].str.contains('one')] # excludes both one and none\n",
    "twoplus.shape # 13 columns, 196530 rows: 1123285 removed, 14.9 % of all data usable\n",
    "# Keep anything noted as 'single nucleotide variant' in 'Type' column\n",
    "snv = twoplus[twoplus['Type'].str.contains('single nucleotide variant')]\n",
    "snv.shape # 174733 rows, or 88.9% of data with two or more stars (13.2% of all data)\n",
    "## 83.2 % of total data, prior to star rating filter, are SNVs\n",
    "# Keep anything with 'Pathogenic' in the 'ClinicalSignificance' column\n",
    "patho = snv[snv['ClinicalSignificance'].str.contains('Pathogenic')]\n",
    "patho.shape # 20040 rows, or 10.2% of data with two or more stars that are snvs (1.5% of all data)\n",
    "## 12.8 % of total data, prior to star rating filter, are Pathogenic"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
